### DISCUSSION TITLE
Introduce high-level interface for string escaping?

### DISCUSSION THREAD

**Discussion by dvarrazzo** [psycopg/psycopg#420] [Ideas]
`psycopg.pq.Escaping` provides access to libpq escaping functions such as [`PQescapeIdentifier()`](https://www.postgresql.org/docs/current/libpq-exec.html#LIBPQ-PQESCAPEIDENTIFIER). However their interface is in bytes and they return memory views, so they are awkward to use.

We could provide higher level functions:

- `escape_identifier()`
- `escape_literal()`
- `escape_string()`
- `escape_bytea()`

to wrap them at higher level (using the connection encoding to encode the input and decode the output).

Issues:

- Some libpq functions require a connection too as argument (identifier, literal). Others (string, bytea) come in two versions: a better one taking the connection too, a deprecated one that doesn't. 
  - Do we keep them asymmetric?
  - Or do we only expose the version taking the connection?
  - Or do we implement a hacked up, buggy version of escaping identifiers and literal not requiring a connection, and we regard it as a bad idea later?
    - However, we could limit these function to take only ascii characters, and we can add tests to make sure that they are escaped correctly, whereas if users have unicode identifiers they would require to pass a connection too.

- Where to we put these functions? I wouldn't like to have them as connection methods, because it feels very bloaty. Some people might want to use these functions because they want to generate off-line sql script and they expect to not need a connection to do that.

  - The `psycopg.sql` method exposes a [`quote()`](https://www.psycopg.org/psycopg3/docs/api/sql.html#psycopg.sql.quote) function which works at even higher level (it also adapts the Python value). Maybe that's the right place where to add them?
  - Or do we say "let's not do this, and force people to use the `psycopg.sql` objects?

**Comment by dlax** [comment]
> Some libpq functions require a connection too as argument (identifier, literal). Others (string, bytea) come in two versions: a better one taking the connection too, a deprecated one that doesn't.

I am in favour of exposing a consistent interface for all functions, that'd be the one requiring a connection. I don't see any real value in exposing deprecated functions; and reading the warnings in libpq documentation doesn't really make me want to use them...

 > Where to we put these functions? I wouldn't like to have them as connection methods, because it feels very bloaty.

Agreed about not putting them as methods (one reason for something to be method is when the code uses the internal state of the object, which would not be here). The `psycopg.sql` module seems like a reasonable place for this.

> Or do we say "let's not do this, and force people to use the psycopg.sql objects?

Seems reasonable too, until somebody comes up with a blocking issue.
